File: src/learning_service/scripts/add_server.py
#!/usr/bin/env python3
# coding:utf-8

import rospy

from learning_service.srv import Add, AddResponse


def addCallback(req):
    """定义回调函数

    Args:
        req (Add.request): 表示Add.srv中的request, 内部具体内容为在客户端中输入的两个参数.

    Returns:
        AddResponse: 返回Add.srv的response, 即输入的和.
    """
    rospy.loginfo(
        "The request and result is: %s + %s = %s", req.a, req.b, (req.a + req.b)
    )
    return AddResponse(req.a + req.b)


def add_server():
    # 初始化add_server节点, 创建名为s的服务端实例, 创建名为add_service的服务,
    # 并将其连接到回调函数addCallback, 该服务的类型为Add.
    rospy.init_node("add_server")
    s = rospy.Service("add_service", Add, addCallback)
    rospy.loginfo("Ready to add two number.")
    rospy.spin()


if __name__ == "__main__":
    add_server()

File: src/learning_service/src/add_server.cpp
#include <ros/ros.h>
#include <learning_service/Add.h>

/**
 * @brief 服务端连接的回调函数
 *
 * @param req learning_service::Add::Request类型的常指针输入, 对应.srv文件的request.
 * @param res earning_service::Add::Response类型的常指针输出, 对应.srv文件的response.
 * @return true 表示回调函数执行成功.
 * @return false
 */
bool addCallback(learning_service::Add::Request &req, learning_service::Add::Response &res)
{
    res.sum = req.a + req.b;
    ROS_INFO("request: x=%0.2f, y=%0.2f", req.a, req.b);
    ROS_INFO("The result is: %0.2f", res.sum);
    return true;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "add_server");
    ros::NodeHandle nh;
    ros::ServiceServer add_service = nh.advertiseService("add_service", addCallback);
    ROS_INFO("Ready to add two number.");
    ros::spin();
    return 0;
}